pkg/ruleengine/v1/r0011_unexpected_network_traffic.go

```diff
@@ -0,0 +1,176 @@
+package ruleengine
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"slices"
+	"strings"
+
+	apitypes "github.com/armosec/armoapi-go/armotypes"
+	"github.com/goradd/maps"
+	"github.com/kubescape/node-agent/pkg/objectcache"
+	"github.com/kubescape/node-agent/pkg/ruleengine"
+	"github.com/kubescape/node-agent/pkg/utils"
+
+	tracernetworktype "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/network/types"
+)
+
+const (
+	R0011ID   = "R0011"
+	R0011Name = "Unexpected network traffic"
+)
+
+var R0011UnexpectedNetworkTrafficRuleDescriptor = RuleDescriptor{
+	ID:          R0011ID,
+	Name:        R0011Name,
+	Description: "Detecting unexpected network traffic that is not whitelisted by application profile.",
+	Tags:        []string{"dns", "whitelisted", "network"},
+	Priority:    RulePriorityMed,
+	Requirements: &RuleRequirements{
+		EventTypes: []utils.EventType{utils.NetworkEventType},
+	},
+	RuleCreationFunc: func() ruleengine.RuleEvaluator {
+		return CreateRuleR0011UnexpectedNetworkTraffic()
+	},
+}
+var _ ruleengine.RuleEvaluator = (*R0011UnexpectedNetworkTraffic)(nil)
+
+type R0011UnexpectedNetworkTraffic struct {
+	BaseRule
+	alertedAdresses maps.SafeMap[string, bool]
+}
+
+func CreateRuleR0011UnexpectedNetworkTraffic() *R0011UnexpectedNetworkTraffic {
+	return &R0011UnexpectedNetworkTraffic{}
+}
+
+func (rule *R0011UnexpectedNetworkTraffic) Name() string {
+	return R0011Name
+}
+func (rule *R0011UnexpectedNetworkTraffic) ID() string {
+	return R0011ID
+}
+
+func (rule *R0011UnexpectedNetworkTraffic) DeleteRule() {
+}
+
+func (rule *R0011UnexpectedNetworkTraffic) handleNetworkEvent(networkEvent *tracernetworktype.Event, objCache objectcache.ObjectCache) ruleengine.RuleFailure {
+	// Check if we already alerted on this address.
+	if ok := rule.alertedAdresses.Has(networkEvent.DstEndpoint.Addr); ok {
+		return nil
+	}
+
+	// Check if the network event is outgoing and the destination is not a private IP.
+	if networkEvent.PktType == "OUTGOING" && !isPrivateIP(networkEvent.DstEndpoint.Addr) {
+		nn := objCache.NetworkNeighborhoodCache().GetNetworkNeighborhood(networkEvent.Runtime.ContainerID)
+		if nn == nil {
+			return nil
+		}
+
+		nnContainer, err := getContainerFromNetworkNeighborhood(nn, networkEvent.GetContainer())
+		if err != nil {
+			return nil
+		}
+
+		domain := objCache.DnsCache().ResolveIpToDomain(networkEvent.DstEndpoint.Addr)
+
+		// Check if the address is in the egress list and isn't in cluster.
+		for _, egress := range nnContainer.Egress {
+			if egress.IPAddress == networkEvent.DstEndpoint.Addr {
+				return nil
+			}
+
+			// Check if we seen this dns name before and it's in-cluster address and in the egress list.
+			if domain != "" && (strings.HasSuffix(domain, "svc.cluster.local.") || slices.Contains(egress.DNSNames, domain)) {
+				return nil
+			}
+		}
+
+		// Alert on the address.
+		rule.alertedAdresses.Set(networkEvent.DstEndpoint.Addr, true)
```

Comment:
Shouldn't this happen per pod/container?

Reply:
The rules are at the container level all ready no?

Reply:
my bad